src/core/coap/coap.cpp

```diff
@@ -456,15 +628,465 @@ void CoapBase::DequeueMessage(Message &aMessage)
+
+    SuccessOrExit(error = SendMessage(*request, aMessageInfo, TxParameters::GetDefault(),
+                                      aCoapMetadata.mResponseHandler, aCoapMetadata.mResponseContext,
+                                      aCoapMetadata.mBlockwiseTransmitHook, aCoapMetadata.mBlockwiseReceiveHook));
+
+exit:
+    if (error != OT_ERROR_NONE && request != NULL)
+    {
+        request->Free();
+    }
+
+    if (optionBuf != NULL)
+    {
+        free(optionBuf);
+        optionBuf = NULL;
+    }
+
+    return error;
+}
+
+otError CoapBase::SendNextBlock2Request(Message &               aRequest,
+                                        Message &               aMessage,
+                                        const Ip6::MessageInfo &aMessageInfo,
+                                        const Metadata &        aCoapMetadata,
+                                        uint32_t                aTotalLength,
+                                        bool                    aBeginBlock1Transfer)
+{
+    otError        error                                        = OT_ERROR_NONE;
+    Message *      request                                      = NULL;
+    uint8_t        buf[OPENTHREAD_CONFIG_COAP_MAX_BLOCK_LENGTH] = {0};
+    uint16_t       bufLen                                       = OPENTHREAD_CONFIG_COAP_MAX_BLOCK_LENGTH;
+    uint8_t *      optionBuf                                    = NULL;
+    OptionIterator iterator;
+    bool           isOptionSet = false;
+
+    SuccessOrExit(error = aMessage.ReadBlockOptionValues(OT_COAP_OPTION_BLOCK2));
+
+    // Check payload and block length
+    VerifyOrExit((aMessage.GetLength() - aMessage.GetOffset()) <=
+                     otCoapBlockSizeFromExponent(aMessage.GetBlockWiseBlockSize()),
+                 error = OT_ERROR_NO_BUFS);
+
+    // Read and then forward payload to receive hook function
+    bufLen = aMessage.Read(aMessage.GetOffset(), aMessage.GetLength() - aMessage.GetOffset(), buf);
+    SuccessOrExit(
+        error = aCoapMetadata.mBlockwiseReceiveHook(aCoapMetadata.mResponseContext, buf,
+                                                    otCoapBlockSizeFromExponent(aMessage.GetBlockWiseBlockSize()) *
+                                                        aMessage.GetBlockWiseBlockNumber(),
+                                                    bufLen, aMessage.IsMoreBlocksFlagSet(), aTotalLength));
+
+    // CoAP Block-Wise Transfer continues
+    otLogInfoCoap("Received Block2 Nr. %d , Size: %d bytes, More Blocks Flag: %d", aMessage.GetBlockWiseBlockNumber(),
+                  otCoapBlockSizeFromExponent(aMessage.GetBlockWiseBlockSize()), aMessage.IsMoreBlocksFlagSet());
+
+    // Conclude block-wise transfer if last block has been received
+    if (!aMessage.IsMoreBlocksFlagSet())
+    {
+        FinalizeCoapTransaction(aRequest, aCoapMetadata, &aMessage, &aMessageInfo, OT_ERROR_NONE);
+        ExitNow();
+    }
+
+    // Init request for next block
```

Comment:
Does this duplicate a bunch of code from `SendNextBlock1Request()`? If so, can we try to avoid the duplication?